File: frontend/src/types/appointment.ts
// ── Enums / Literals ──────────────────────────────────────────────────────

export type AppointmentType = 'scheduled' | 'walk-in' | 'emergency' | 'follow_up';
export type VisitType = 'new_visit' | 'follow_up' | 'referral' | 'emergency';
export type AppointmentStatus = 'scheduled' | 'pending' | 'confirmed' | 'in-progress' | 'completed' | 'cancelled' | 'no-show' | 'rescheduled';
export type Priority = 'normal' | 'urgent' | 'emergency';

// ── Doctor Schedule ──────────────────────────────────────────────────────

export interface DoctorSchedule {
  id: string;
  doctor_id: string;
  day_of_week: number;  // 0=Sun
  start_time: string;
  end_time: string;
  slot_duration_minutes: number;
  max_patients: number | null;
  break_start_time: string | null;
  break_end_time: string | null;
  is_active: boolean;
  effective_from: string | null;
  effective_to: string | null;
  created_at: string;
  updated_at: string;
}

export interface DoctorScheduleCreate {
  day_of_week: number;
  start_time: string;
  end_time: string;
  slot_duration_minutes?: number;
  max_patients?: number;
  break_start_time?: string;
  break_end_time?: string;
  is_active?: boolean;
}

// ── Doctor Leave (replaces Blocked Period) ───────────────────────────────

export interface DoctorLeave {
  id: string;
  doctor_id: string;
  leave_date: string;
  leave_type: string;
  reason: string | null;
  is_half_day: boolean;
  half_day_period: string | null;
  approved_by: string | null;
  created_at: string;
}

export interface DoctorLeaveCreate {
  doctor_id: string;
  leave_date: string;
  leave_type?: string;
  reason?: string;
}

// ── Appointment ──────────────────────────────────────────────────────────

export interface Appointment {
  id: string;
  hospital_id: string | null;
  appointment_number: string;
  patient_id: string;
  doctor_id: string | null;
  department_id: string | null;
  appointment_type: AppointmentType;
  visit_type: string | null;
  appointment_date: string;
  start_time: string | null;
  end_time: string | null;
  status: AppointmentStatus;
  priority: string | null;
  chief_complaint: string | null;
  consultation_fee: number | null;
  cancel_reason: string | null;
  reschedule_reason: string | null;
  reschedule_count: number;
  check_in_at: string | null;
  consultation_start_at: string | null;
  consultation_end_at: string | null;
  notes: string | null;
  created_by: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  // Enriched fields
  patient_name?: string | null;
  doctor_name?: string | null;
  department_name?: string | null;
  patient_reference_number?: string | null;
  doctor_specialization?: string | null;
}

export interface AppointmentCreate {
  patient_id: string;
  doctor_id?: string | null;
  department_id?: string | null;
  appointment_type?: string;
  visit_type?: string;
  appointment_date: string;
  start_time?: string | null;
  chief_complaint?: string;
  priority?: string;
  consultation_fee?: number | null;
}

export interface AppointmentUpdate {
  doctor_id?: string | null;
  department_id?: string | null;
  appointment_date?: string;
  start_time?: string | null;
  visit_type?: string;
  chief_complaint?: string;
  priority?: string;
  consultation_fee?: number;
  notes?: string;
  doctor_notes?: string;
}

// ── Walk-in ──────────────────────────────────────────────────────────────

export interface WalkInRegister {
  patient_id: string;
  doctor_id: string;
  department_id?: string | null;
  chief_complaint?: string;
  priority?: string;
  consultation_fee?: number | null;
}

export interface QueueItem {
  queue_id: string;
  appointment_id: string;
  queue_number: number;
  position: number;
  status: string;
  priority: string;
  patient_name: string | null;
  patient_id: string | null;
  patient_reference_number: string | null;
  patient_phone: string | null;
  patient_gender: string | null;
  patient_date_of_birth: string | null;
  patient_age: number | null;
  patient_blood_group: string | null;
  patient_email: string | null;
  patient_known_allergies: string | null;
  patient_chronic_conditions: string | null;
  patient_emergency_contact_name: string | null;
  patient_emergency_contact_phone: string | null;
  patient_emergency_contact_relation: string | null;
  doctor_id: string | null;
  doctor_name: string | null;
  chief_complaint: string | null;
  check_in_at: string | null;
  called_at: string | null;
  consultation_start_at: string | null;
  consultation_end_at: string | null;
}

export interface QueueStatus {
  doctor_id: string | null;
  queue_date: string;
  total_waiting: number;
  total_in_progress: number;
  total_completed: number;
  items: QueueItem[];
}

// ── Time Slot ────────────────────────────────────────────────────────────

export interface TimeSlot {
  time: string;
  available: boolean;
  current_bookings: number;
  max_bookings: number;
}

export interface AvailableSlots {
  doctor_id: string;
  date: string;
  slots: TimeSlot[];
}

// ── Stats ────────────────────────────────────────────────────────────────

export interface AppointmentStats {
  total_appointments: number;
  total_scheduled: number;
  total_walk_ins: number;
  total_completed: number;
  total_cancelled: number;
  total_no_shows: number;
  total_pending: number;
  completion_rate: number;
  cancellation_rate: number;
  no_show_rate: number;
  average_wait_time: number;
}

// ── Doctor (for dropdowns) ───────────────────────────────────────────────

export interface DoctorOption {
  doctor_id: string;
  user_id: string;
  name: string;
  specialization: string | null;
  department_id: string | null;
  consultation_fee: number | null;
}

// ── Generic Paginated ────────────────────────────────────────────────────

export interface PaginatedResponse<T> {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  data: T[];
}

// ── Enhanced Stats ───────────────────────────────────────────────────────

export interface DoctorUtilization {
  doctor_id: string;
  doctor_name: string;
  department: string | null;
  total_appointments: number;
  completed: number;
  cancelled: number;
  no_shows: number;
  utilization_rate: number;
}

export interface DepartmentBreakdown {
  department: string;
  total: number;
  completed: number;
  cancelled: number;
  no_shows: number;
}

export interface TrendDataPoint {
  date: string;
  total: number;
  completed: number;
  cancelled: number;
  no_shows: number;
}

export interface PeakTimeSlot {
  hour: number;
  count: number;
  label: string;
}

export interface CancellationReason {
  reason: string;
  count: number;
}

export interface EnhancedAppointmentStats extends AppointmentStats {
  doctor_utilization: DoctorUtilization[];
  department_breakdown: DepartmentBreakdown[];
  trends: TrendDataPoint[];
  peak_times: PeakTimeSlot[];
  cancellation_reasons: CancellationReason[];
}

// ── Waitlist ─────────────────────────────────────────────────────────────

export type WaitlistStatus = 'waiting' | 'notified' | 'booked' | 'cancelled' | 'expired';

export interface WaitlistEntry {
  id: string;
  hospital_id: string;
  patient_id: string;
  doctor_id: string;
  department_id: string | null;
  preferred_date: string;
  preferred_time: string | null;
  appointment_type: string;
  priority: string;
  chief_complaint: string | null;
  reason: string | null;
  status: WaitlistStatus;
  position: number;
  booked_appointment_id: string | null;
  notified_at: string | null;
  expires_at: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  // Enriched
  patient_name: string | null;
  patient_reference_number: string | null;
  patient_phone: string | null;
  doctor_name: string | null;
  doctor_specialization: string | null;
}

export interface WaitlistCreate {
  patient_id: string;
  doctor_id: string;
  department_id?: string | null;
  preferred_date: string;
  preferred_time?: string | null;
  appointment_type?: string;
  priority?: string;
  chief_complaint?: string;
  reason?: string;
}

export interface PaginatedWaitlist {
  total: number;
  page: number;
  limit: number;
  data: WaitlistEntry[];
}

export interface WaitlistStats {
  total_waiting: number;
  total_booked: number;
  total_cancelled: number;
  total_expired: number;
  total: number;
}

export interface WalkInResponse {
  // Normal walk-in fields
  id?: string;
  appointment_number?: string;
  queue_number?: number | null;
  queue_position?: number | null;
  // Waitlist auto-add fields
  waitlisted?: boolean;
  message?: string;
  waitlist_entry?: WaitlistEntry;
  [key: string]: unknown;
}

File: backend/app/schemas/appointment.py
from pydantic import BaseModel, Field, field_validator, model_validator, ConfigDict
from typing import Optional, Any
from datetime import date, time, datetime
from decimal import Decimal


VALID_APPOINTMENT_TYPES = ["scheduled", "walk-in"]
VALID_APPOINTMENT_STATUSES = [
    "scheduled", "pending", "confirmed", "in-progress", "completed",
    "cancelled", "no-show", "rescheduled",
]
VALID_PRIORITY_LEVELS = ["normal", "urgent", "emergency"]
VALID_LEAVE_TYPES = ["personal", "sick", "holiday", "conference", "other"]
WEEKDAY_NAMES = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]


# ---- Helper: ORM -> dict with stringified UUIDs ----

def _orm_to_dict(data: Any) -> Any:
    """Convert SQLAlchemy model instance to dict with all UUIDs as strings."""
    if hasattr(data, "__table__"):
        d = {}
        for col in data.__table__.columns:
            val = getattr(data, col.name)
            # Convert UUID fields to str
            if hasattr(val, "hex"):  # uuid.UUID
                val = str(val)
            d[col.name] = val
        return d
    if isinstance(data, dict):
        return data
    return data


# ---- Doctor Schedule Schemas ----

class DoctorScheduleBase(BaseModel):
    day_of_week: int = Field(..., ge=0, le=6, description="0=Sunday, 1=Monday ... 6=Saturday")
    start_time: time
    end_time: time
    slot_duration_minutes: int = Field(default=15, ge=5, le=120)
    max_patients: int = Field(default=20, ge=1, le=100)
    break_start_time: Optional[time] = None
    break_end_time: Optional[time] = None
    is_active: bool = True
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None

    @model_validator(mode="after")
    def default_effective_from(self):
        if self.effective_from is None:
            self.effective_from = date.today()
        return self

    @field_validator("end_time")
    @classmethod
    def validate_end_after_start(cls, v: time, info) -> time:
        start = info.data.get("start_time")
        if start and v <= start:
            raise ValueError("end_time must be after start_time")
        return v


class DoctorScheduleCreate(DoctorScheduleBase):
    pass


class DoctorScheduleUpdate(BaseModel):
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    slot_duration_minutes: Optional[int] = Field(default=None, ge=5, le=120)
    max_patients: Optional[int] = Field(default=None, ge=1, le=10)
    break_start_time: Optional[time] = None
    break_end_time: Optional[time] = None
    is_active: Optional[bool] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None


class DoctorScheduleResponse(BaseModel):
    id: str
    doctor_id: str
    day_of_week: int
    start_time: time
    end_time: time
    slot_duration_minutes: int
    max_patients: int
    break_start_time: Optional[time] = None
    break_end_time: Optional[time] = None
    is_active: bool
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class DoctorScheduleBulkCreate(BaseModel):
    doctor_id: str
    schedules: list[DoctorScheduleCreate]


# ---- Doctor Leave Schemas ----

class DoctorLeaveCreate(BaseModel):
    doctor_id: str
    leave_date: date
    leave_type: str = Field(default="full_day", description="full_day, morning, or afternoon")
    reason: Optional[str] = Field(None, max_length=255)

    @field_validator("leave_type")
    @classmethod
    def validate_leave_type(cls, v: str) -> str:
        valid = ["full_day", "morning", "afternoon"]
        if v not in valid:
            raise ValueError(f"Must be one of: {', '.join(valid)}")
        return v


class DoctorLeaveResponse(BaseModel):
    id: str
    doctor_id: str
    leave_date: date
    leave_type: str
    reason: Optional[str] = None
    status: Optional[str] = None
    approved_by: Optional[str] = None
    created_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


# ---- Appointment Schemas ----

class AppointmentCreate(BaseModel):
    patient_id: str
    doctor_id: Optional[str] = None
    department_id: Optional[str] = None
    appointment_type: str = Field(default="scheduled")
    visit_type: Optional[str] = None
    appointment_date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    chief_complaint: Optional[str] = None
    priority: Optional[str] = Field(default="normal")
    consultation_fee: Optional[Decimal] = None

    @field_validator("appointment_type")
    @classmethod
    def validate_appointment_type(cls, v: str) -> str:
        if v not in VALID_APPOINTMENT_TYPES:
            raise ValueError(f"Must be one of: {', '.join(VALID_APPOINTMENT_TYPES)}")
        return v

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: Optional[str]) -> Optional[str]:
        if v and v not in VALID_PRIORITY_LEVELS:
            raise ValueError(f"Must be one of: {', '.join(VALID_PRIORITY_LEVELS)}")
        return v


class AppointmentUpdate(BaseModel):
    doctor_id: Optional[str] = None
    appointment_date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    chief_complaint: Optional[str] = None
    priority: Optional[str] = None
    consultation_fee: Optional[Decimal] = None
    visit_type: Optional[str] = None
    department_id: Optional[str] = None


class AppointmentStatusUpdate(BaseModel):
    status: str

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        if v not in VALID_APPOINTMENT_STATUSES:
            raise ValueError(f"Must be one of: {', '.join(VALID_APPOINTMENT_STATUSES)}")
        return v


class AppointmentReschedule(BaseModel):
    new_date: date
    new_time: Optional[time] = None
    reason: Optional[str] = None


class AppointmentResponse(BaseModel):
    id: str
    appointment_number: str
    patient_id: str
    doctor_id: Optional[str] = None
    hospital_id: Optional[str] = None
    department_id: Optional[str] = None
    appointment_type: str
    visit_type: Optional[str] = None
    appointment_date: date
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    status: str
    priority: Optional[str] = "normal"
    chief_complaint: Optional[str] = None
    consultation_fee: Optional[Decimal] = None
    cancel_reason: Optional[str] = None
    reschedule_reason: Optional[str] = None
    reschedule_count: int = 0
    created_by: Optional[str] = None
    check_in_at: Optional[datetime] = None
    consultation_start_at: Optional[datetime] = None
    consultation_end_at: Optional[datetime] = None
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime
    patient_name: Optional[str] = None
    doctor_name: Optional[str] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class AppointmentListItem(BaseModel):
    id: str
    appointment_number: str
    patient_id: str
    doctor_id: Optional[str] = None
    appointment_type: str
    visit_type: Optional[str] = None
    appointment_date: date
    start_time: Optional[time] = None
    status: str
    priority: Optional[str] = "normal"
    chief_complaint: Optional[str] = None
    consultation_fee: Optional[Decimal] = None
    notes: Optional[str] = None
    created_at: datetime
    patient_name: Optional[str] = None
    doctor_name: Optional[str] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return data
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class PaginatedAppointmentResponse(BaseModel):
    total: int
    page: int
    limit: int
    total_pages: int
    data: list[AppointmentListItem]


# ---- Walk-in Schemas ----

class WalkInRegister(BaseModel):
    patient_id: str
    doctor_id: Optional[str] = None
    chief_complaint: Optional[str] = None
    priority: str = Field(default="normal")
    consultation_fee: Optional[Decimal] = None

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: str) -> str:
        if v not in VALID_PRIORITY_LEVELS:
            raise ValueError(f"Must be one of: {', '.join(VALID_PRIORITY_LEVELS)}")
        return v


class WalkInAssignDoctor(BaseModel):
    doctor_id: str


class QueueStatus(BaseModel):
    total_waiting: int
    total_in_progress: int
    total_completed_today: int
    average_wait_time: int
    queue: list[AppointmentListItem]


# ---- Settings Schemas (Hospital-level) ----

class HospitalSettingsResponse(BaseModel):
    id: str
    hospital_id: str
    appointment_slot_duration: int = 30
    appointment_buffer_minutes: int = 5
    allow_walk_ins: bool = True
    max_advance_booking_days: int = 30
    cancellation_policy_hours: int = 24
    enable_auto_reminders: bool = True
    enable_sms_notifications: bool = False
    enable_email_notifications: bool = True
    working_hours_start: Optional[time] = None
    working_hours_end: Optional[time] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class HospitalSettingsUpdate(BaseModel):
    # Appointment Configuration
    appointment_slot_duration_minutes: Optional[int] = Field(None, ge=5, le=120)
    appointment_buffer_minutes: Optional[int] = Field(None, ge=0, le=60)
    max_daily_appointments_per_doctor: Optional[int] = Field(None, ge=1, le=100)
    consultation_fee_default: Optional[str] = Field(None, max_length=20)
    follow_up_validity_days: Optional[int] = Field(None, ge=1, le=365)
    # Legacy aliases (for backward compat)
    appointment_slot_duration: Optional[int] = Field(None, ge=5, le=120)
    allow_walk_ins: Optional[bool] = None
    # Notifications
    enable_sms_notifications: Optional[bool] = None
    enable_email_notifications: Optional[bool] = None
    enable_whatsapp_notifications: Optional[bool] = None
    # ID Sequences
    hospital_code: Optional[str] = Field(None, min_length=2, max_length=2, pattern=r'^[A-Z]{2}$')
    patient_id_start_number: Optional[int] = Field(None, ge=1)
    staff_id_start_number: Optional[int] = Field(None, ge=1)
    invoice_prefix: Optional[str] = Field(None, max_length=10)
    prescription_prefix: Optional[str] = Field(None, max_length=10)
    # Compliance
    data_retention_years: Optional[int] = Field(None, ge=1, le=99)

    @field_validator("hospital_code")
    @classmethod
    def validate_hospital_code(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.isupper():
            v = v.upper()
        return v

    @field_validator("consultation_fee_default")
    @classmethod
    def validate_consultation_fee(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            try:
                float(v)  # Must be numeric string
            except ValueError:
                raise ValueError("Consultation fee must be a valid numeric value")
        return v


# ---- Stats / Report Schemas ----

class AppointmentStats(BaseModel):
    total_appointments: int = 0
    total_scheduled: int = 0
    total_walk_ins: int = 0
    total_completed: int = 0
    total_cancelled: int = 0
    total_no_shows: int = 0
    total_pending: int = 0
    completion_rate: float = 0.0
    cancellation_rate: float = 0.0
    no_show_rate: float = 0.0
    average_wait_time: float = 0.0


class EnhancedAppointmentStats(AppointmentStats):
    """Extended stats with per-doctor and per-department breakdowns."""
    doctor_stats: list[dict] = []
    department_stats: list[dict] = []
    daily_trends: list[dict] = []
    peak_hours: list[dict] = []
    cancellation_reasons: list[dict] = []


class AppointmentSettingResponse(BaseModel):
    key: str
    value: str
    description: Optional[str] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class AppointmentSettingUpdate(BaseModel):
    setting_value: str


class TimeSlot(BaseModel):
    time: time
    available: bool
    current_bookings: int
    max_bookings: int


class AvailableSlotsResponse(BaseModel):
    doctor_id: str
    date: date
    slots: list[TimeSlot]


# ---- Waitlist Schemas ----

VALID_WAITLIST_STATUSES = ["waiting", "notified", "booked", "cancelled", "expired"]


class WaitlistCreate(BaseModel):
    patient_id: str
    doctor_id: str
    department_id: Optional[str] = None
    preferred_date: date
    preferred_time: Optional[time] = None
    appointment_type: str = Field(default="walk-in")
    priority: str = Field(default="normal")
    chief_complaint: Optional[str] = None
    reason: Optional[str] = None

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: str) -> str:
        if v not in VALID_PRIORITY_LEVELS:
            raise ValueError(f"Must be one of: {', '.join(VALID_PRIORITY_LEVELS)}")
        return v


class WaitlistUpdate(BaseModel):
    status: Optional[str] = None
    preferred_date: Optional[date] = None
    preferred_time: Optional[time] = None
    priority: Optional[str] = None
    reason: Optional[str] = None

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v and v not in VALID_WAITLIST_STATUSES:
            raise ValueError(f"Must be one of: {', '.join(VALID_WAITLIST_STATUSES)}")
        return v


class WaitlistResponse(BaseModel):
    id: str
    hospital_id: str
    patient_id: str
    doctor_id: str
    department_id: Optional[str] = None
    preferred_date: date
    preferred_time: Optional[time] = None
    appointment_type: str
    priority: str
    chief_complaint: Optional[str] = None
    reason: Optional[str] = None
    status: str
    position: int
    booked_appointment_id: Optional[str] = None
    notified_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    # Enriched fields (set by router)
    patient_name: Optional[str] = None
    patient_reference_number: Optional[str] = None
    patient_phone: Optional[str] = None
    doctor_name: Optional[str] = None
    doctor_specialization: Optional[str] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class PaginatedWaitlistResponse(BaseModel):
    total: int
    page: int
    limit: int
    data: list[WaitlistResponse]
